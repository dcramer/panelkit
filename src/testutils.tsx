import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
