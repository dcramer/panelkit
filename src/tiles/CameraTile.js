import React from "react";
import PropTypes from "prop-types";

import Thumbnail from "../components/Thumbnail";
import Tile, { TileProps } from "./Tile";
import CameraModal from "../components/CameraModal";

export default class CameraTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
    refreshInterval: PropTypes.number,
  };

  static defaultProps = {
    refreshInterval: 3000,
  };

  getInitialState() {
    return {
      loading: true,
      result: null,
      showModal: false,
    };
  }

  componentDidMount() {
    super.componentDidMount();
    this.refreshCameraImage();
  }

  componentWillUnmount() {
    super.componentWillUnmount();
    if (this._timer) clearTimeout(this._timer);
  }

  getWatchedEntityIds() {
    return [];
  }

  refreshCameraImage = () => {
    return this.props.hass
      .fetchCameraThumbnail(this.props.entityId)
      .then(({ result }) => {
        this.setState({
          result,
          loading: false,
        });
      })
      .then(() => {
        this._timer = setTimeout(
          this.refreshCameraImage,
          this.props.refreshInterval
        );
      });
  };

  onTouch = () => {
    if (this.state.showModal) return;
    this.setState({ showModal: true });
  };

  closeModal = () => {
    this.setState({ showModal: false });
  };

  renderBody() {
    return (
      <CameraModal
        hass={this.props.hass}
        entityId={this.props.entityId}
        isOpen={this.state.showModal}
        onRequestClose={this.closeModal}
        cameraList={this.props.cameraList}
      />
    );
  }

  renderStatus() {}

  renderCover() {
    const { loading, result } = this.state;
    if (loading) return <div>loading camera</div>;
    return <Thumbnail result={result} alt={this.props.entityId} />;
  }
}
