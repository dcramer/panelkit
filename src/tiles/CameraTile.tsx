import React from "react";

import Thumbnail from "../components/Thumbnail";
import Tile, { TileProps, TileState } from "./Tile";
import CameraModal, { CameraModalProps } from "../modals/CameraModal";

type CameraTileProps = TileProps & {
  entityId: string;
  refreshInterval?: number;
};

type CameraTileState = TileState & {
  loading: boolean;
  cameraUrl: string | null;
};

export default class CameraTile extends Tile<CameraTileProps, CameraTileState> {
  readonly state: CameraTileState = {
    loading: true,
    cameraUrl: null,
    modalIsOpen: false,
    isLoading: false,
  };

  static defaultProps = {
    refreshInterval: 3000,
  };

  private _timer?: number;

  componentDidMount() {
    super.componentDidMount();
    this.refreshCameraImage();
  }

  componentWillUnmount() {
    super.componentWillUnmount();
    if (this._timer) clearTimeout(this._timer);
  }

  getWatchedEntityIds() {
    return [];
  }

  async getCameraThumbnailUrl() {
    const { hass, entityId } = this.props;
    const result = await hass.signPath(`/api/camera_proxy/${entityId}`);
    return result;
  }

  refreshCameraImage = () => {
    this.getCameraThumbnailUrl()
      .then((cameraUrl) => {
        this.setState({
          cameraUrl,
          loading: false,
        });
      })
      .then(() => {
        this._timer = setTimeout(
          this.refreshCameraImage,
          this.props.refreshInterval
        );
      });
  };

  onTouch = () => {
    this.openModal();
  };

  renderModal(params: CameraModalProps) {
    return <CameraModal {...params} />;
  }

  renderStatus() {}

  renderCover() {
    const { loading, cameraUrl } = this.state;
    if (loading || !cameraUrl) return <div>loading camera</div>;
    return (
      <Thumbnail
        url={this.props.hass.buildUrl(cameraUrl)}
        alt={this.props.entityId}
      />
    );
  }
}
