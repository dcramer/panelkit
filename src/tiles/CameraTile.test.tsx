import { renderTile } from "../testutils";
import CameraTile from "./CameraTile";

test("renders", async () => {
  const { container, getByText } = renderTile(CameraTile, {
    entity_id: "camera.test",
    state: "idle",
    attributes: {
      friendly_name: "Test Camera",
    },
  });

  expect(getByText(/Test Camera/i)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
