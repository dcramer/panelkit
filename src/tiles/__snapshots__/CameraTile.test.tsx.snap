// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="tile state-idle touchable"
  data-testid="tile"
  style="cursor: pointer;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-cover"
    >
      <div>
        loading camera
      </div>
    </div>
    <div
      class="tile-body"
    />
    <div
      class="tile-title"
    >
      Test Camera
    </div>
  </div>
</div>
`;
